xt_imager.py

```diff
@@ -0,0 +1,259 @@
+import pathlib
+import serial
+import logging
+import argparse
+import traceback
+from typing import List
+from string import printable
+import tftpy
+import threading
+import lzma
+
+logging.basicConfig(level=logging.INFO, format="[%(levelname)s] %(message)s")
+log = logging.getLogger(__name__)
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description='Flash image files through u-boot and tftp')
+
+    parser.add_argument(
+        'image',
+        type=pathlib.Path,
+        help='Name of image file to flash')
+
+    parser.add_argument(
+        '-s',
+        '--serial',
+        help='Serial console to use')
+
+    parser.add_argument(
+        '-t',
+        '--tftp',
+        nargs='?',
+        type=str,
+        default=None,
+        const="AUTO",
+        help="Use external TFTP server or start our own")
+
+    parser.add_argument(
+        '--serverip',
+        help='IP of the host that will be used TFTP transfer.')
+
+    parser.add_argument(
+        '--ipaddr',
+        help='IP of the board that will be used TFTP transfer.')
+
+    args = parser.parse_args()
+
+    tftp_root = os.getcwd()
+
+    if args.tftp == "AUTO":
+        log.info(f"Starting our TFTP server...")
+
+        tftpsrv = PYTFTPServer(tftp_root)
+        TFTP_srv_thread = threading.Thread(name="TFTP Server thread", target=tftpsrv.start_tftp_server)
+        TFTP_srv_thread.start()
+
+    elif os.path.isdir(args.tftp):
+        # use external path
+        tftp_root = args.tftp
+        log.info(f"Use external TFTP root {tftp_root}")
+    else:
+        raise Exception("-t parameter is not external TFTP root.")
+
+    do_flash_image(args, tftp_root)
+
+    if args.tftp == "AUTO":
+        log.info("Stopping our TFTP server")
+        tftpsrv.stop_tftp_server()
+        TFTP_srv_thread.join()
+
+
+class PYTFTPServer(object):
+    def __init__(self, folder):
+        self.tftp_server = tftpy.TftpServer(folder)
+
+    def start_tftp_server(self):
+        # listen to all interfaces and port 69
+        self.tftp_server.listen()
+
+    def stop_tftp_server(self):
+        self.tftp_server.stop()
+
+
+def do_flash_image(args, tftp_root):
+
+    log.info(args.image)
+
+    conn = open_connection(args)
+
+    uboot_propmt = "=>"
+
+    # Send 'CR', and check for one of the possible options:
+    # - uboot_prompt appears, if u-boot console is already active
+    # - u-boot is just starting, so we will get "Hit any key.."
+    log.info('Waiting for u-boot prompt...')
+    conn_send(conn, "\r")
+    conn_wait_for_any(conn, [uboot_propmt, "Hit any key to stop autoboot:"])
+    # we got "Hit any key", so let's stop the boot
+    conn_send(conn, "\r")
+    conn_wait_for(conn, uboot_propmt)
+
+    image_size = os.path.getsize(args.image)
+
+    ###########
+    base_addr = 0x0
+    mmc_device = 1
+    mmc_part = 0
+    mmc_block_size = 512
+    ###########
+
+    chunk_filename = "chunk.bin"
+    chunk_size_in_bytes = 20*1024*1024
+
+    use_lzma = False
```

Comment:
not required , the following code should work 
if str(args.image).endswith(".xz"):
        f = lzma.open(args.image)
        image_size = 0
    else:
        f = open(args.image, "rb")
all the rest is the same

Reply:
In the initial stages of development, it was just like you proposed.
But I decided to use a dedicated variable because I do not feel comfortable with string manipulations inside the loop, which (manipulation) is not required, and can be 'cached' inside bool.

Reply:
then use just use_lxma